cmd/snap/cmd_run.go

```diff
@@ -227,6 +230,141 @@ func isReexeced() bool {
+	}
+
+	xauthPath := osGetenv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return "", nil
+	}
+
+	fin, err := os.Open(xauthPath)
+	if err != nil {
+		return "", err
+	}
+	defer fin.Close()
+
+	// Abs() also calls Clean(); see https://golang.org/pkg/path/filepath/#Abs
+	xauthPathAbs, err := filepath.Abs(fin.Name())
+	if err != nil {
+		return "", nil
+	}
+
+	// Remove all symlinks from path
+	xauthPathCan, err := filepath.EvalSymlinks(xauthPathAbs)
+	if err != nil {
+		return "", nil
+	}
+
+	// Ensure the XAUTHORITY env is not abused by checking that
+	// it point to exactly the file we just opened (no symlinks,
+	// no funny "../.." etc)
+	if fin.Name() != xauthPathCan {
+		logger.Noticef("WARNING: %s != %s\n", fin.Name(), xauthPathCan)
+		return "", nil
+	}
+
+	// Only do the migration from /tmp since the real /tmp is not visible for snaps
+	if !strings.HasPrefix(fin.Name(), "/tmp/") {
+		return "", nil
+	}
+
+	// We are performing a Stat() here to make sure that the user can't
+	// steal another user's Xauthority file. Note that while Stat() uses
+	// fstat() on the file descriptor created during Open(), the file might
+	// have changed ownership between the Open() and the Stat(). That's ok
+	// because we aren't trying to block access that the user already has:
+	// if the user has the privileges to chown another user's Xauthority
+	// file, we won't block that since the user can just steal it without
+	// having to use snap run. This code is just to ensure that a user who
+	// doesn't have those privileges can't steal the file via snap run
+	// (also note that the (potentially untrusted) snap isn't running yet).
+	fi, err := fin.Stat()
+	if err != nil {
+		return "", err
+	}
+	sys := fi.Sys()
+	if sys == nil {
+		return "", fmt.Errorf(i18n.G("Can't validate file owner"))
+	}
+	// cheap comparison as the current uid is only available as a string
+	// but it is better to convert the uid from the stat result to a
+	// string than a string into a number.
+	if fmt.Sprintf("%d", sys.(*syscall.Stat_t).Uid) != u.Uid {
+		return "", fmt.Errorf(i18n.G("Xauthority file isn't owned by the current user %s"), u.Uid)
+	}
+
+	targetPath := filepath.Join(baseTargetDir, ".Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// otherwise we can continue using the existing Xauthority file.
+	// This is ok to do here because we aren't trying to protect against
+	// the user changing the Xauthority file in XDG_RUNTIME_DIR outside
+	// of snapd.
+	if osutil.FileExists(targetPath) {
+		var fout *os.File
+		if fout, err = os.Open(targetPath); err != nil {
+			return "", err
+		}
+		if osutil.FileStreamsEqual(fin, fout) {
+			fout.Close()
+			return targetPath, nil
+		}
+
+		fout.Close()
+		if err := os.Remove(targetPath); err != nil {
+			logger.Noticef("WARNING: failed to remove existing file %s", targetPath)
```

Comment:
Here we apparently warn and then return the same error, which is then warned again from the call site.